### PR TITLE
Add JWT authentication

### PR DESCRIPTION
**Pull request by zengyan-amazon**
*Issue #, if available:*

*Description of changes:* Add JWT (JsonWebToken) authentication.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: server/auth/types/jwt/jwt_auth.ts [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+      if (authHeaderValue) {
+        const header: any = {};
+        header[this.authHeaderName] = authHeaderValue;
+        return tookit.authenticated({
+          requestHeaders: header,
+        });
+      }
+    }
+
+    // not credetail in query parameter, header, or cookie,
```
Comment (yuxitang-amzn) [inline server/auth/types/jwt/jwt_auth.ts:105]: typo

File: server/auth/types/jwt/jwt_auth.ts [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+      token = (request.url.query as ParsedUrlQuery)[urlParamName] as string;
+      if (token) {
+        return token;
+      }
+    }
+
+    // fallback to HTTP header
+    const authHeaderValue = request.headers[this.authHeaderName];
+    if (authHeaderValue && !Array.isArray(authHeaderValue)) {
+      // authHeaderValue should not be array, to ensure client doesn't pass multiple auth header
```
Comment (yuxitang-amzn) [inline server/auth/types/jwt/jwt_auth.ts:139]: Could we move the comment above the corresponding code?

File: server/plugin.ts [hunk 3]
```diff
@@ -110,6 +111,15 @@ export class OpendistroSecurityPlugin
         this.logger
       );
       core.http.registerAuth(auth.authHandler);
+    } else if (config.auth.type == 'jwt') {
```
Comment (yuxitang-amzn) [inline server/plugin.ts:114]: Is it better to have an enum for all the auth types?

File: server/auth/types/jwt/jwt_auth.ts (hunk 1 above)
Comment (zengyan-amazon) [reply server/auth/types/jwt/jwt_auth.ts:105]: fixed

File: server/auth/types/jwt/jwt_auth.ts (hunk 2 above)
Comment (zengyan-amazon) [reply server/auth/types/jwt/jwt_auth.ts:139]: moved

File: server/plugin.ts (hunk 3 above)
Comment (zengyan-amazon) [reply server/plugin.ts:114]: agreed, will refactor later in a separate PR

File: server/auth/types/jwt/jwt_auth.ts [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    private readonly sessionStorageFactory: SessionStorageFactory<SecuritySessionCookie>,
+    private readonly router: IRouter,
+    private readonly esClient: IClusterClient,
+    private readonly coreSetup: CoreSetup
+  ) {
+    this.authHeaderName = this.config.jwt?.header.toLowerCase() || 'authorization';
+    this.securityClient = new SecurityClient(esClient);
+  }
+
+  authHandler: AuthenticationHandler = async (request, response, tookit) => {
```
Comment (tianleh) [inline server/auth/types/jwt/jwt_auth.ts:48]: nonblocking comment: probably break this function into 3 subfunctions for better readability, authenticate by jwt, authenticate by cookie, or redirect.

**Pull request merged**